### ISSUE TITLE
Dynamic targets: drop targets that are NULL from further processing?

### ISSUE THREAD

**Issue by mike-lawrence** [wlandau/targets#199]
## Prework

* [x] Read and agree to the [code of conduct](https://github.com/wlandau/targets/blob/main/CODE_OF_CONDUCT.md) and [contributing guidelines](https://github.com/wlandau/targets/blob/main/CONTRIBUTING.md).
* [x] If there is [already a relevant issue](https://github.com/wlandau/targets/issues), whether open or closed, comment on the existing thread instead of posting a new issue.

## Question

I'm using dynamic targets to process lots of small pieces of data, and sometimes a piece might be corrupted in a way that I can detect so when this happens I simply have my function return NULL for that piece. However, subsequent functions I apply will throw an error if the input is NULL, so I have to get rid of the NULL pieces somehow. I've worked out that creating a non-dynamic target that is is the result of calling `purrr:compact()` will achieve this, but it takes a bunch of extra time. Is there any more efficient way to drop NULL pieces that I'm missing? Here's a brief example pipeline:

```r
tar_pipeline(
	... #geting the filenames here

	#dynamic target to get each data chunk
	, tar_target(
		name = raw_data
		, command = read_data(files) #sometimes this returns NULL
		, pattern = map(files)
		, iteration = 'list' #so we can use purr::compact() next
	)

	#non-dynamic target to toss NA chunks
	, tar_target(
		name = compacted_raw_data
		, command = purrr::compact(raw_data)
	)

	#dynamic target again
	, tar_target(
		name = lowpassed
		, command = apply_lowpass(compacted_raw_data, filter_order = 1)
		, pattern = map(compacted_raw_data)
	)
)
```

**Comment by mattwarkentin** [comment]
Hi @mike-lawrence,

Step 2 can be avoided by making sure Step 3 can throw an error safely when provided with `NULL` inputs. One of the simplest ways to do this would be to wrap the function used in Step 3 with `purrr::possibly()`.

For example,
```r
apply_lowpass_possibly <- purrr::possibly(apply_lowpass, otherwise = NULL)
```
Then when you want to use `apply_lowpass()` in a situation where it might thrown an error, use `apply_lowpass_possibly()` instead. That way it will run even when it hits an error, and you get control over what value is returned when an error is faced simply by modifying the `otherwise` argument. `purrr::safely()` and `purrr::quietly()` perform similar but different jobs compared to `purrr::possibly()`. Feel free to use whichever works best. The other added benefit is that the target value(s) for Step 1 and Step 3 will contain the same number of items, which may be valuable for whatever you want to do downstream.

These `purrr` functions are basically just simple ways to do conditional handling, instead of using something like `tryCatch()` which is more flexible, but more involved to use.

**Comment by mike-lawrence** [comment]
Thanks for your response. If I understand correctly, if I have lots of steps after step 3, each one is going to have to have a way of handling the error condition output of the prior, right? It's that kind of re-coding of every single step that I'd like to avoid ideally.

**Comment by mattwarkentin** [comment]
Yes, that's correct. And I agree, having to make sure every downstream function is able to handle errors is not ideal. Step 2 really isn't a bad way to flatten out the `NULL`s, I don't think.

How many list-items are there in `raw_data`? I wouldn't think Step 2 would take that much time to remove empty fields.

**Comment by wlandau** [comment]
Thanks, @mattwarkentin. I agree with your assessment: `purrr` functions to control side effects, taking out step 2, custom code to handle `NULL`s in step 3, and possibly `tar_option_set(error = "continue")`.

I would add that if you can return a value from step 1 to make the code just work, you can avoid much of the headache later on. Are you working with data frames? If it helps, I would consider returning a zero-row data frame with the exact column names that your downstream code expects.

This next point may not be tangibly useful here, but it is worth noting that we are getting deeper into functional programming territory here. This struggle is exactly the sort of problem that [monads](https://en.wikipedia.org/wiki/Monad_(functional_programming) are designed to solve, particularly the `Maybe` monad. A monad is just a box that houses the return value and side effects, and it helps avoid having to write messy logic that depends on the value itself. `targets` borrows this construct internally (though the infrastructure as a whole is far more object-oriented than functional). Related: https://github.com/arendsee/rmonad.

**Comment by mike-lawrence** [comment]
Huh, I might have happened on a bug while attempting to use the code in my original post. When it gets to the `compacted_raw_data` target, it pins a core at 100% indefinitely. There's a file in `_targets/objects` called `compacted_raw_data` and it reads in fine, though it's a single tibble with the rows from all the chunks concatenated rather than a list of tibbles as I expected. I've attempted to create a minimal reprex, but I'm having trouble making one that actually produces the same hang behaviour. I've attached the most [elaborated attempt](https://github.com/wlandau/targets/files/5412531/failed_reprex.zip), as well as [a zip](https://github.com/wlandau/targets/files/5412553/_targets.zip) of the `_targets` dir from the real pipeline that hangs. Let me know if I can do anything else to help figure out what's going awry here.

**Comment by mattwarkentin** [comment]
Is it possible you didn't have `iteration = 'list'` for Step 1? If you forgot this line then by default `targets` will use `vctrs::vec_c()` to concatenate the items in Step 1, and the effect is row binding of data frames.
```r
# This returns a row-bound data.frame excluding NULL
vctrs::vec_c(iris, NULL, iris)
```

**Comment by mike-lawrence** [comment]
In the version that has the hang, I indeed had `iteration = 'list'`.

**Comment by mattwarkentin** [comment]
This target in your `failed_reprex.R` script needs `iteration = "list"` too.
```r
tar_target(
  name = data, 
  command = read_data(files), 
  pattern = map(files)
)
```

**Comment by mike-lawrence** [comment]
Ah, gotcha. The user guide says that the `iteration` type propogates to all downstream targets, so I didn't think the `data` target needed it specified, but I can see that maybe since `files` is not a mapped target, it doesn't actually use the `iteration` argument and thus it's not propogated to `data` even though that is a mapped target. When fixed, `compacted_data` indeed is a list of tibbles as I originally expected. This also fixes the full version that hangs! Still odd that I can't create a reprex of the hang itself, but thanks for your help in getting past it.

**Comment by mattwarkentin** [comment]
No problem! Happy to help.

Though I think `targets` behaviour is as expected in your case. @wlandau can correct me if I'm wrong, but the propagation refers to how a target itself is treated when used in downstream targets. 

So `files` is set to `iteration = "list"` so targets know when `files` is mapped over in downstream targets, the items should be accessed by `list[[i]]`. However, you get to choose a new iteration method for `data`, and in this case the default is `vctrs::vec_c()`. If you want `data` to be treated as a `list` then you need to specify that also. Thus if `data` is mapped over in a later step, again the items would be accessed via `list[[i]]`. Since you just used `data` as a full list in the last step, there is no slicing or aggregating going on, so `iteration` doesn't matter.

**Comment by mattwarkentin** [comment]
If the documentation does use the word __"propagate"__ then maybe it needs to be changed to something like "the `iteration` type is used in all downstream targets" or something...

**Comment by mike-lawrence** [comment]
It doesn't, sorry, that was just my misinterpretation. In 5.4.2 (list iteration), it says "List iteration splits and aggregates targets as simple lists. If target x has "list" iteration, all branches of downstream patterns will get x[[1]], x[[2]], and so on. "

**Comment by wlandau** [comment]
> Though I think targets behaviour is as expected in your case. @wlandau can correct me if I'm wrong, but the propagation refers to how a target itself is treated when used in downstream targets, and not that it forces that same iteration selection on downstream targets that depend on said target.

That's right. `iteration` is about how the current target presents itself to neighbors immediately downstream. It does not control how those neighbors present themselves in turn.
